Add blog post on React Native and JHipster (#1024)
oktadeveloper/okta-blog: _source/_posts/2022-09-12-react-native-spring-boot-jhipster.adoc

---
layout: blog_post
title: "Integrate React Native and Spring Boot Securely"
author: matt-raible
by: advocate
communities: [mobile,java,javascript]
description: ""
tags: []
tweets:
- ""
image:
type: conversion
github: https://github.com/oktadev/auth0-react-native-jhipster-example
---
:page-liquid:
:toc: macro
:experimental:
:commandkey: &#8984;

// If this post is not targeting keywords, "Get started with React Native for JHipster" is probably a better title.
//   - 55: Integrate React Native and Spring Boot Securely
//   - 57: React Native Authentication with OAuth and Spring Boot
//   - 60: Secure React Native and Spring Boot with OpenID Connect
//   - 61: Build React Native Apps Quickly with JHipster

> //   - 65: Build Secure React Native Apps Quickly with JHipster

Author: We could also match the [Ionic blueprint post](https://developer.okta.com/blog/2022/05/12/ionic-angular-jhipster) from May. That is, use `react-native-jhipster` for the URL and `Build Secure React Native Apps with JHipster` (62) for the title. The Ionic blueprint post hasn't done very well in terms of popularity. Since launch it only has 960 sessions. However, 577 of those are new users.

// Keywords:
//   - react native security: 320, react native spring boot: 140, react native oauth: 260
//   - react native authentication: 1300, react native tutorial: 18,100
//
// https://auth0.com/blog/get-started-auth0-authentication-react-native-android/ is #5 for react native authentication
// https://developer.okta.com/code/react-native/ is #18 for react native authentication

React Native is a mobile app framework from Facebook. It allows you to quickly develop apps using React's API and deploy them to iOS and Android. It allows you to quickly refresh the apps when you make changes and generally offers a pleasant experience for web developers.

> https://necolas.github.io/react-native-web/[React Native for Web] is a recent addition to the React Native family. It allows you to run your app in a browser and enjoy the developer tools it offers. It's optimized for mobile and uses the same components as React Native, so it looks good too!

Reviewer: ```suggestion
https://necolas.github.io/react-native-web/[React Native for Web] is a recent addition to the React Native family. It allows you to run your app in a browser and enjoy the browser's built-in dev tools. It's optimized for mobile and uses the same components as React Native, so it looks good too!
```
"enjoy the developer tools it offers" - does this "it" refer to browser dev tools? If yes, it should be explicit...  otherwise, it's still a lot of "it"s...

In this tutorial, I'll show you how to create a React Native application that deploys to the web, iOS, and Android. You'll configure it to use OpenID Connect (OIDC) for authentication and use OAuth 2.0 access tokens to talk securely to a Spring Boot API.

[.red]#todo: add architecture diagram#

**Prerequisites:**

* https://adoptopenjdk.net/[Java 11+]
* https://nodejs.org[Node 16+]

If you're on Windows, you may need to install the https://docs.microsoft.com/en-us/windows/wsl/about[Windows Subsystem for Linux] for some commands to work.

I recommend using SDKMAN to manage your OpenJDK installations. Just run `sdk install java 11.0.2-open` to install Java 11 and `sdk install java 17-open` for Java 17.

toc::[]

> == Quick Apps with JHipster

Reviewer: ```suggestion
== Quick apps with JHipster
```
sentence case

== JHipster blueprints for more power!

JHipster added a **blueprints** feature several years ago. It allows you, as a developer, to create a project that overrides the default behavior of JHipster. This feature has led to a https://www.jhipster.tech/modules/official-blueprints/[thriving ecosystem of blueprints], from Kotlin to Micronaut to .NET Core to NestJS to Svelte to Ionic and even React Native.

Today, I'll show you how to use https://github.com/jhipster/generator-jhipster-react-native[JHipster's React Native blueprint] to build a Flickr clone. The backend will be powered by the Spring Boot app created in https://auth0.com/blog/full-stack-java-with-react-spring-boot-and-jhipster/[Full Stack Java with React, Spring Boot, and JHipster] tutorial. The React Native app's screens will be generated from the same JDL (JHipster Domain Language) file that created the backend entities.

I'm excited to show you how to use JHipster React Native because its https://github.com/jhipster/generator-jhipster-react-native/releases/tag/v4.3.0[4.3.0 release] upgrades it Expo 46, React Native 0.69.5, and React 18. Let's giddyup! 🤠

> == Build a Photo Gallery with React and Spring Boot

Reviewer: ```suggestion
== Build a photo gallery with React and Spring Boot
```

[source,shell]
----
cd backend
docker-compose -f src/main/docker/keycloak.yml up -d
----

// _If you'd rather skip this part, I'll show you how to configure and use Auth0 in a minute._

.Apple Silicon and Keycloak
****
If you're using Apple Silicon (aka M1 or M2), Keycloak will fail to start because its Docker image wasn't available for an M1 when the backend project was created. To fix it, run the script below.

[source,shell]
----
VERSION=15.0.2 # Keycloak version specified in keycloak.yml
cd /tmp
git clone git@github.com:keycloak/keycloak-containers.git
cd keycloak-containers/server
git checkout $VERSION
docker build -t "jboss/keycloak:${VERSION}" .
docker build -t "quay.io/keycloak/keycloak:${VERSION}" .
----
****

Then, start the backend using `./mvnw` and open your favorite browser to `\http://localhost:8080`. You should be able to log in with `admin/admin` and upload photos. They'll be displayed in a nice grid, and you can click each photo to zoom in.

image::{% asset_path blog/full-stack-java/photo-gallery.jpg %}[alt=Gallery with Photos,width=800,align=center]

Now, let's create a React Native app that talks to the same API.

> == Generate a React Native App

Reviewer: ```suggestion
== Generate a React Native app
```

Install React Native JHipster and the Expo CLI:

[source,shell]
----
npm install -g generator-jhipster-react-native expo-cli
----

Create a directory for your React Native app:

[source,shell]
----
take mobile
----

Run the following command to use the React Native blueprint to create an app.

[source,shell]
----
jhipster --blueprints react-native # you can also use `rnhipster`
----

When prompted, use the following values:

[cols="1,1"]
|===
|Prompt |Answer

|What do you want to name your React Native application?
|`Flickr2`

|Enter the directory where your JHipster app is located:
|`../backend`

|Do you want to enable end-to-end tests with Detox?
|`No`
|===

Next, generate screens based on the entities in the backend project. Press **a** (for **all**) when prompted to overwrite files.

[source,shell]
----
rnhipster jdl ../backend/flickr2.jdl
----

In the backend project, change its `src/main/resources/config/application-dev.yml` to allow `\http://localhost:19006` for CORS (cross-origin resource sharing):

[source,yaml]
----
cors:
  allowed-origins: 'http://localhost:19006,...'
----

https://expo.dev/signup[Sign up for an Expo account] and take note of your username.

> Login to http://localhost:9080/auth/admin/[Keycloak] (with `admin/admin` as credentials). Navigate to **Clients** > **web_app** and add `\https://auth.expo.io/@<your-expo-username>/Flickr2` as a Valid Redirect URI. **Save** your changes.

Reviewer: ```suggestion
Log in to http://localhost:9080/auth/admin/[Keycloak] (with `admin/admin` as credentials). Navigate to **Clients** > **web_app** and add `\https://auth.expo.io/@<your-expo-username>/Flickr2` as a Valid Redirect URI. **Save** your changes.
```

For the Android emulator to communicate with your API and Keycloak, you'll need to add some port mappings. The command worked if `8080` and `9080` are printed to your terminal.

[source,shell]
----
adb reverse tcp:8080 tcp:8080 && adb reverse tcp:9080 tcp:9080
----

CAUTION: If you get a `command not found` error, see https://stackoverflow.com/questions/10303639/adb-command-not-found[this Stack Overflow Q & A] to solve it. I used `echo export "PATH=~/Library/Android/sdk/platform-tools:$PATH" >> ~/.zshrc` on my Mac.

To reload your app, hit kbd:[r] twice with a focus on the Android emulator.

[cols="a,a,a", frame=none, grid=none]
|===
| image::{% asset_path blog/react-native-jhipster/android-expo-prompt.png %}[alt=Android Expo prompt,width=300,align=center]
| image::{% asset_path blog/react-native-jhipster/android-keycloak-login.png %}[alt=Android Keycloak login,width=300,align=center]
| image::{% asset_path blog/react-native-jhipster/android-keycloak-signed-in.png %}[alt=Android Keycloak login,width=300,align=center]
|===

> == Use Auth0 for Identity

Reviewer: ```suggestion
== Use Auth0 for identity
```

Author: I actually did this one on purpose because I liked the look of Identity in uppercase. I'll commit your suggestion through.

    api.idToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
    api.accessToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
  }
}
----

This code adds the user's roles to a custom claim (prefixed with `\https://www.jhipster.tech/roles`). This claim is mapped to Spring Security authorities in `SecurityUtils.java` on the backend.

Select **Deploy** and drag the `Add Roles` action to your Login flow. Create a `backend/.auth0.env` file and populate it with your Auth0 settings.

[source,shell]
----
export SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_OIDC_ISSUER_URI=https://<your-auth0-domain>/
export SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_ID=<your-client-id>
export SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_SECRET=<your-client-secret>
export JHIPSTER_SECURITY_OAUTH2_AUDIENCE=https://<your-auth0-domain>/api/v2/
----

NOTE: Want to have all these steps automated for you? Watch https://github.com/auth0/auth0-cli/issues/351[issue #351] in the Auth0 CLI project.

Stop your backend app with kbd:[Ctrl+C] and run the following commands to configure Spring Security to use Auth0.

[source,shell]
----
source .auth0.env
./mvnw
----

> === Create a Native OIDC App for React Native

Reviewer: ```suggestion
=== Create a native OIDC app for React Native
```

----
http://localhost:19006/,https://auth.expo.io/@<your-expo-username>/Flickr2
----

Configure Allowed Logout URLs:

----
http://localhost:19006,https://auth.expo.io/@<your-expo-username>/Flickr2
----

And, set the Allowed Origins (CORS):

----
http://localhost:19006,http://localhost
----

The second value is the origin header that Android sends. Copy the client ID to `app/config/app-config.js` and update the `audience` in `app/modules/login/login.utils.ts`:

[source,js]
----
audience: 'https://<your-auth0-domain>/api/v2/',
----

Restart your React Native app and log in with Auth0!

[cols="a,a", frame=none, grid=none]
|===
| image::{% asset_path blog/react-native-jhipster/auth0-login.png %}[alt=Auth0 login,width=300,align=center]
| image::{% asset_path blog/react-native-jhipster/auth0-login-success.png %}[alt=Auth0 login success,width=300,align=center]
|===

It works on Android too. 🥳

[cols="a,a", frame=none, grid=none]
|===
| image::{% asset_path blog/react-native-jhipster/android-auth0-login.png %}[alt=Android Auth0 login,width=300,align=center]
| image::{% asset_path blog/react-native-jhipster/android-signed-in.png %}[alt=Android Auth0 login,width=300,align=center]
|===

> === Use Okta for Identity

Reviewer: ```suggestion
=== Use Okta for identity
```

[source,shell]
----
cd backend
docker-compose -f src/main/docker/keycloak.yml up -d
----

// _If you'd rather skip this part, I'll show you how to configure and use Auth0 in a minute._

.Apple Silicon and Keycloak
****
If you're using Apple Silicon (aka M1 or M2), Keycloak will fail to start because its Docker image wasn't available for an M1 when the backend project was created. To fix it, run the script below.

[source,shell]
----
VERSION=15.0.2 # Keycloak version specified in keycloak.yml
cd /tmp
git clone git@github.com:keycloak/keycloak-containers.git
cd keycloak-containers/server
git checkout $VERSION
docker build -t "jboss/keycloak:${VERSION}" .
docker build -t "quay.io/keycloak/keycloak:${VERSION}" .
----
****

Then, start the backend using `./mvnw` and open your favorite browser to `\http://localhost:8080`. You should be able to log in with `admin/admin` and upload photos. They'll be displayed in a nice grid, and you can click each photo to zoom in.

image::{% asset_path blog/full-stack-java/photo-gallery.jpg %}[alt=Gallery with Photos,width=800,align=center]

> Now, let's create a React Native app that talks to the same API.

Reviewer 2: ```suggestion
Now let's create a React Native app that talks to the same API.
```

| image::{% asset_path blog/react-native-jhipster/react-native-edit-photo.jpg %}[alt=Edit photo,width=400,align=center,link={% asset_path blog/react-native-jhipster/react-native-edit-photo.jpg %}]
|===

=== Test on iOS

To see your React Native app running on iOS, press **i** in the window you ran `npm start` from. You will need to be on a Mac with Xcode installed for this to work.

[cols="a,a", frame=none, grid=none]
|===
| image::{% asset_path blog/react-native-jhipster/react-native-ios.png %}[alt=React Native app on iOS,width=350,align=center]
| image::{% asset_path blog/react-native-jhipster/react-native-ios-photo.jpg %}[alt=React Native photo on iOS,width=350,align=center]
|===

TIP: You can reload your app in Simulator using kbd:[{commandKey} + R].

=== Test on Android

To see your React Native app running on Android, press **a** in the window you ran `npm start` from. You will need Android Studio and an AVD (Android Virtual Device) running. I tested on a Pixel 5 with API 31 (Android 12.0).

> For the Android emulator to communicate with your API and Keycloak, you'll need to add some port mappings. The command worked if `8080` and `9080` are printed to your terminal.

Reviewer: ```suggestion
For the Android emulator to communicate with your API and Keycloak, you'll need to add some port mappings. You'll know the command worked if `8080` and `9080` are printed to your terminal.
```
adds context.
